src/compiler/checker.ts

```diff
@@ -12585,6 +12585,19 @@ namespace ts {
                         else if (grandParent.kind === SyntaxKind.TypeParameter && grandParent.parent.kind === SyntaxKind.MappedType) {
                             inferences = append(inferences, keyofConstraintType);
                         }
+                        // When an 'infer T' declaration is the template of a mapped type, and that mapped type if the extends
+                        // clause of a conditional whose check type is also a mapped type, give it the constraint of the template
+                        // of the check type's mapped type

```

Comment:
> give it the constraint of the template of the check type's mapped type

I think this phrasing is ambiguous, and I want to nitpick it just to figure out whether I’m understanding the goal. It could be interpreted in one of two ways:

1. Give `infer T` a constraint that _is the constraint of_ the template type on the check side of the conditional type
2. Give `infer T` a constraint that _is_ the template type on the check side of the conditional type

Reading the code and the test case, I _think_ what you mean is (2), because given

```ts
type KeysWithoutStringIndex<T> =
     { [K in keyof T]: string extends K ? never : K } extends { [_ in keyof T]: infer U }
//                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ “template of the check type’s mapped type”
     ? U
     : never
```

I don’t know what it means for `string extends K ? never : K` to _have_ a constraint in this case, and it looks like the type you’re appending to `inferences` is just the type of `string extends K ? never : K` instantiated with `K` → `keyof T`. Did I understand that right?

Reply:
Yes on both fronts. That constraint will then later be executed and become a single type.